python/cudf/cudf/tests/test_groupby.py

```diff
@@ -376,6 +377,81 @@ def emulate(df):
     assert_groupby_results_equal(expect, got)
 
 
+@pytest.fixture(scope="module")
+def groupby_jit_data():
+    np.random.seed(0)
+    df = DataFrame()
+    nelem = 20
+    df["key1"] = np.random.randint(0, 3, nelem)
+    df["key2"] = np.random.randint(0, 2, nelem)
+    df["val1"] = np.random.random(nelem)
+    df["val2"] = np.random.random(nelem)
+    return df
+
+
+def run_groupby_apply_jit_test(data, func, keys, *args):
+    expect_groupby_obj = data.to_pandas().groupby(keys, as_index=False)
+    got_groupby_obj = data.groupby(keys)
+
+    cudf_jit_result = got_groupby_obj.apply(func, *args, engine="jit")
+    pandas_result = expect_groupby_obj.apply(func, *args)
+    # compare cuDF jit to pandas
+
+    assert_groupby_results_equal(cudf_jit_result, pandas_result)
+
+
+@pytest.mark.parametrize("dtype", SUPPORTED_GROUPBY_NUMPY_TYPES)
+@pytest.mark.parametrize(
+    "func", ["min", "max", "sum", "mean", "var", "std", "idxmin", "idxmax"]
+)
+def test_groupby_apply_jit_reductions(func, groupby_jit_data, dtype):
+    # dynamically generate to avoid pickling error
```

Comment:
You're right to ask about this. It has to do with the way we determine the cache key for the incoming udf [here.](https://github.com/rapidsai/cudf/blob/branch-23.02/python/cudf/cudf/utils/cudautils.py#L204)

Ideally I'd like this to read:

```python
def f(df):
    return getattr(df['val1'], func)()
```

But this will throw at the linked line with
```
AttributeError: Can't pickle local object 'test_groupby_apply_jit_reductions.<locals>.func'
```
I believe this comes down to the `getattr` call as the culprit being a "child function" that can't be pickled. Rather than rework caching for all UDFs as part of this PR I decided to generate these tests dynamically.

Reply:
Got it. I thought it had to do with Numba execution, so I was surprised that pickling was involved at any point. I didn't realize it was related to caching. Can we clarify that the pickling issue is cache-related, and then file an issue and link it here?